File: sdk/js/src/kernel.ts
// Copyright 2026 Ojima Abraham
// SPDX-License-Identifier: Apache-2.0

/// Kernel compilation and launch for the WAVE JavaScript/TypeScript SDK.

import { DeviceInfo } from "./device";
import { WaveArray } from "./array";
import { compileKernel, launchEmulator } from "./runtime";

/** Supported source languages. */
export type Language = "python" | "rust" | "cpp" | "typescript";

/** A compiled WAVE kernel ready for launch. */
export class CompiledKernel {
  private source: string;
  private language: Language;
  private wbin: Buffer | null = null;

  constructor(source: string, language: Language = "typescript") {
    this.source = source;
    this.language = language;
  }

  private async ensureCompiled(): Promise<Buffer> {
    if (this.wbin === null) {
      this.wbin = await compileKernel(this.source, this.language);
    }
    return this.wbin;
  }

  /** Launch the kernel on the given device. */
  async launch(
    dev: DeviceInfo,
    buffers: WaveArray[],
    scalars: number[],
    grid: [number, number, number] = [1, 1, 1],
    workgroup: [number, number, number] = [256, 1, 1],
  ): Promise<void> {
    const wbin = await this.ensureCompiled();

    if (dev.vendor === "emulator") {
      await launchEmulator(wbin, buffers, scalars, grid, workgroup);
    } else {
      throw new Error(
        `Direct ${dev.vendor} launch not yet implemented in JS SDK. ` +
          "Use the emulator or the Rust SDK.",
      );
    }
  }
}

/** Create a compiled kernel from source code. */
export function kernel(
  source: string,
  language: Language = "typescript",
): CompiledKernel {
  return new CompiledKernel(source, language);
}

File: sdk/python/src/wave_gpu/kernel.py
# Copyright 2026 Ojima Abraham
# SPDX-License-Identifier: Apache-2.0

"""Kernel decorator and compilation for the WAVE Python SDK."""

import inspect
import textwrap
from typing import Any, Callable, List, Optional

from .array import WaveArray
from .runtime import CompiledKernel


class KernelWrapper:
    """Wraps a Python function as a WAVE GPU kernel."""

    def __init__(self, func: Callable[..., Any]) -> None:
        self._func = func
        self._source = textwrap.dedent(inspect.getsource(func))
        self._compiled: Optional[CompiledKernel] = None

    def _ensure_compiled(self) -> CompiledKernel:
        if self._compiled is None:
            self._compiled = CompiledKernel(self._source, "python")
        return self._compiled

    def __call__(self, *args: Any, **kwargs: Any) -> None:
        """Launch the kernel with the given arguments.

        Buffer arguments (WaveArray) are passed as device buffers.
        Scalar arguments (int) are passed as kernel parameters.
        """
        grid = kwargs.get("grid", (1, 1, 1))
        workgroup = kwargs.get("workgroup", (256, 1, 1))

        buffers: List[WaveArray] = []
        scalars: List[int] = []

        for arg in args:
            if isinstance(arg, WaveArray):
                buffers.append(arg)
            elif isinstance(arg, (int, float)):
                scalars.append(int(arg))
            else:
                raise TypeError(
                    f"Unsupported argument type: {type(arg).__name__}. "
                    "Expected WaveArray or int."
                )

        compiled = self._ensure_compiled()

        n_threads = max(len(b) for b in buffers) if buffers else 1
        if grid == (1, 1, 1) and workgroup == (256, 1, 1):
            wg_size = min(256, n_threads)
            n_groups = (n_threads + wg_size - 1) // wg_size
            grid = (n_groups, 1, 1)
            workgroup = (wg_size, 1, 1)

        compiled.launch(buffers, scalars, grid, workgroup)


def kernel(func: Callable[..., Any]) -> KernelWrapper:
    """Decorator that marks a Python function as a WAVE GPU kernel."""
    return KernelWrapper(func)
